server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -514,4 +506,437 @@ boolean containsBrokenAnalysis(String field) {
         }
         return reloadedAnalyzers;
     }
+
+    /**
+     * An immutable snapshot of the current mapping.
+     */
+    public interface Snapshot {
```

Comment:
On this topic, have you experimented moving the methods that depend on Document mapper to document mapper? QueryShardContext could then have its own internal snapshot of document mapper, or even a direct final reference to it retrieved through MapperService that already exposes it? Maybe I am missing details around this, but I was wondering if this could be a good first step before we streamline using the snapshot more, as the main goal of this change is to make the query phase use the same mappings throughout its execution.

Reply:
>  the snapshot holds methods that don't depend on DocumentMapper, hence are not affected by the volatile reference being swapped. I am referring to indexAnalyzers, parserContext and isMetadataField. These methods don't belong to the snapshot, but I guess we can't take them out unless we make their consumers depend on two things. Would it be worth it trying to address this if possible?

Yeah, I pushed the mutable bits together with the immutable bit because I figure it'd be convenient to depend on one "thing" for the mapping stuff. It feels convenient for `QueryShradContext`.

> On this topic, have you experimented moving the methods that depend on Document mapper to document mapper? QueryShardContext could then have its own internal snapshot of document mapper, or even a direct final reference to it retrieved through MapperService that already exposes it?

I poked at it a bit but didn't like the idea of adding the version to it because we'd have to move the merging stuff into it. And it felt better to leave it in MapperService.